pumpp/task/key.py

```diff
@@ -0,0 +1,438 @@
+    annotation, where the mode is reflected in the 12-D pitch_profile vector.
+
+    Attributes
+    ----------
+    name : str
+        The name of the key trnsformer
+
+    sr : number > 0
+        The sampling rate of audio
+
+    hop_length : int > 0
+        The number of samples between each annotation frame
+
+    sparse : bool
+        If True, tonic value is sparsely encoded as integers in [0, 12].
+        If False, tonic value is densely encoded as 13-dimensional booleans.
+    '''
+    def __init__(self, name='key', sr=22050, hop_length=512, sparse=False):
+        '''Initialize a key task transformer'''
+
+        super(KeyTransformer, self).__init__(name=name,
+                                             namespace='key_mode',
+                                             sr=sr, hop_length=hop_length)
+        self.sparse = sparse
+ 
+        self.register('pitch_profile', [None, 12], np.bool)
+        if self.sparse:
+            self.register('tonic', [None, 1], np.int)
+        else:
+            self.register('tonic', [None, 13], np.bool)
+
+    def _encode_key_str(self, key_str):
+        '''Helper function to go from jams `key_mode` annotation value strings to 12-D 
+        numpy membership vec, representing the pitch profile.
+
+        Parameters
+        ----------
+        key_str : str
+            String in the style of 'key_mode' jams annotation values.
+
+        Returns
+        -------
+        (pitch_profile, tonic) : tuple
+            pitch_profile : np.ndarray, shape = (1, 12), dtype = np.bool
+                a 12-D row vector that's encodes the membership of each pitch class for 
+                a given `key_str`.
+            tonic : int or np.ndarray, shape = (1, 13), dtype = np.bool
+                a int in the range [0, 12] to indicate the pitch class of the tonic. 12
+                being atonal.
+        '''
+        
+        key_str_split = key_str.split(':')
+        
+        # Look at the Tonic first
+        if key_str_split[0] == 'N':
+            tonic = 12
+        else:
+            tonic = note_to_midi(key_str_split[0]) % 12
+
+        # Now look at quality/mode and build pitch_profile
+        # First construct the profile in C for a given mode/quality
+        c_major_profile = np.zeros(12)
+        for pc in C_MAJOR_PITCHES:
+            c_major_profile[pc] = 1
+
+        # When there is no tonal center, pitch profile is all zeros.
+        if tonic == 12:
+            pitch_profile = np.zeros(12, dtype=np.bool)
+        else:
+            # When there is no quality, major assumed.
+            if len(key_str_split) == 1:
+                quality = 'major'
+            else:
+                quality = key_str_split[1]
+
+            if quality in MODES:
+                mode_transpose_int = -1 * C_MAJOR_PITCHES[MODES.index(quality)]
+            elif quality in QUALITY.keys():
+                mode_transpose_int = -1 * QUALITY[quality]
```

Comment:
what if both of these fail?  do we have an `X` class?

Reply:
When I wrote it I didn't consider the `X` class.

This is because I was trying to follow the syntax of jams:
'''The value field is a string matching one of the three following patterns:

N : no key
Ab, A, A#, Bb, ... G# : tonic note, upper case
tonic:MODE where tonic is as described above, and MODE is one of: major, minor, ionian, dorian, phrygian, lydian, mixolydian, aeolian, locrian'''

Do we want to have `N` for no key (silence) and an `X` for keys out of vocab?
I want to say no and keep only `N` for now because to include `X` properly I think JAMS needs a PR as well... Thoughts?

Reply:
On a re-reading of the code I wrote... I see what you are saying now, so disregard my previous comment.
The issue is that when quality is out of vocab, what do we do. It shouldn't just turn the entire label to `N` because we still have a clear tonic. (Consider someone writes `G:pentatonic`)

What should the behavior really be when the function encounters a string that contains legal `tonic` but out of vocab `mode`?

If I have a catch-all for `mode` string like an `X`, that would mean later in the KeyTagTransformer, the vocab will increase by 12 to include `C:X`, `C#:X`, etc.
And the encoded `pitch_profile` should be a 12-d vec with a `1` on the tonic and `0` elsewhere?

Reply:
I think I'm going do the forbidden for now:
If it's out of vocab... then it's Major! :)
It'll also raise a warning to the user.

Reply:
That sounds reasonable.